### PR TITLE
emacsPackages: add updateScript's

### PR DESCRIPTION
**Pull request by AndersonTorres**
## Description of changes

From https://github.com/NixOS/nixpkgs/issues/278925

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/editors/emacs/elisp-packages/manual-packages/wat-mode/default.nix [hunk 1]
```diff
@@ -1,9 +1,8 @@
-# Manually packaged until it is upstreamed to melpa
-# See https://github.com/devonsparks/wat-mode/issues/1
```
Comment (jian-lin) [inline pkgs/applications/editors/emacs/elisp-packages/manual-packages/wat-mode/default.nix:2]: Not sure if we should delete this.  If you really want to do this, please do so in a separate commit.

File: pkgs/applications/editors/emacs/elisp-packages/manual-packages/voicemacs/default.nix [hunk 2]
```diff
@@ -44,6 +45,8 @@ melpaBuild {
     yasnippet
   ];
 
+  passthru.updateScript = unstableGitUpdater { hardcodeZeroVersion = true; };
```
Comment (jian-lin) [inline pkgs/applications/editors/emacs/elisp-packages/manual-packages/voicemacs/default.nix:48]: I suggest that we do not add a update script to packages without a maintainer.  Auto bump PRs need people to test.

File: pkgs/applications/editors/emacs/elisp-packages/manual-packages/copilot/default.nix [hunk 3]
```diff
@@ -3,10 +3,12 @@
   dash,
   editorconfig,
   fetchFromGitHub,
+  melpaBuild,
   nodejs,
   s,
-  melpaBuild,
```
Comment (jian-lin) [inline pkgs/applications/editors/emacs/elisp-packages/manual-packages/copilot/default.nix:8]: Not sure about moving parameters around.  If you really want to do this, please do so in a separate PR.

File: pkgs/applications/editors/emacs/elisp-packages/manual-packages.nix [hunk 4]
```diff
@@ -29,7 +29,9 @@ in
 
   el-easydraw = callPackage ./manual-packages/el-easydraw { };
 
-  elisp-ffi = callPackage ./manual-packages/elisp-ffi { };
+  elisp-ffi = callPackage ./manual-packages/elisp-ffi {
+    inherit (pkgs) libffi pkg-config;
+  };
```
Comment (jian-lin) [inline pkgs/applications/editors/emacs/elisp-packages/manual-packages.nix:36]: What is the motivation?  It works without this change.

File: pkgs/applications/editors/emacs/elisp-packages/manual-packages.nix [hunk 5]
```diff
@@ -17,7 +17,9 @@ in
 
   cask = callPackage ./manual-packages/cask { };
 
-  codeium = callPackage ./manual-packages/codeium { };
+  codeium = callPackage ./manual-packages/codeium {
+    inherit (pkgs) codeium;
```
Comment (jian-lin) [inline pkgs/applications/editors/emacs/elisp-packages/manual-packages.nix:21]: Please factor out this change into a separate commit.

File: pkgs/applications/editors/emacs/elisp-packages/manual-packages/codeium/default.nix [hunk 6]
```diff
@@ -19,17 +20,19 @@ melpaBuild {
 
   patches = [
     (substituteAll {
-      src = ./codeium.el.patch;
-      codeium = "${pkgs.codeium}/bin/codeium_language_server";
+      src = ./0000-set-codeium-command-executable.patch;
```
Comment (jian-lin) [inline pkgs/applications/editors/emacs/elisp-packages/manual-packages/codeium/default.nix:23]: Please factor out this renaming into a separate commit.

File: pkgs/applications/editors/emacs/elisp-packages/manual-packages/codeium/default.nix [hunk 7]
```diff
@@ -19,17 +20,19 @@ melpaBuild {
...
+      codeium = "${lib.getExe' codeium "codeium_language_server"}";
```
Comment (jian-lin) [inline pkgs/applications/editors/emacs/elisp-packages/manual-packages/codeium/default.nix:24]: Not need to do `${}`.

File: pkgs/applications/editors/emacs/elisp-packages/manual-packages/wat-mode/default.nix (hunk 1 above)
Comment (AndersonTorres) [reply pkgs/applications/editors/emacs/elisp-packages/manual-packages/wat-mode/default.nix:2]: https://github.com/NixOS/nixpkgs/pull/329232

File: pkgs/applications/editors/emacs/elisp-packages/manual-packages/codeium/default.nix (hunk 6 above)
Comment (AndersonTorres) [reply pkgs/applications/editors/emacs/elisp-packages/manual-packages/codeium/default.nix:23]: https://github.com/NixOS/nixpkgs/pull/329237

File: pkgs/applications/editors/emacs/elisp-packages/manual-packages/voicemacs/default.nix (hunk 2 above)
Comment (AndersonTorres) [reply pkgs/applications/editors/emacs/elisp-packages/manual-packages/voicemacs/default.nix]: I will adopt them, then.

https://github.com/NixOS/nixpkgs/pull/329248
https://github.com/NixOS/nixpkgs/pull/329249

File: pkgs/applications/editors/emacs/elisp-packages/manual-packages.nix (hunk 4 above)
Comment (AndersonTorres) [reply pkgs/applications/editors/emacs/elisp-packages/manual-packages.nix]: I am being overzealous here.
I prefer to be explicit about which packages came from external `pkgs`, while the elisp ones can be kept implicit.

We do not have a package like `emacsPackages.pkg-config` yet, but better safe than sorry.

Comment (jian-lin) [reply pkgs/applications/editors/emacs/elisp-packages/manual-packages.nix]: Actually, this change is semantically significant.  Before, arguments like `libffi` are from [`pkgs'`][1] which cannot be `override`ed.  With this change applied, arguments like `libffi` are from [`pkgs`][2] which can be `override`ed.  So theoretically this change can cause breakage if users `override` `pkgs` to an attrset not containing needed arguments like `libffi`.

An example of `override` which **does not** overrides `pkgs` can be fount at [emacs-overlay][3].

[1]: https://github.com/NixOS/nixpkgs/blob/7c3a071a4c1cbb4d5628d1abb2cad5fc01898740/pkgs/top-level/emacs-packages.nix#L55
[2]: https://github.com/NixOS/nixpkgs/blob/7c3a071a4c1cbb4d5628d1abb2cad5fc01898740/pkgs/top-level/emacs-packages.nix#L56
[3]: https://github.com/nix-community/emacs-overlay/blob/7164cf7ca4318ffad776682ef8cf8aa9e24a1216/overlays/package.nix#L28

Comment (jian-lin) [reply pkgs/applications/editors/emacs/elisp-packages/manual-packages.nix]: After checking some other package sets in Nixpkgs, such as `qt6Packages`, I find that the convention is to use the old way.

Comment (AndersonTorres) [reply pkgs/applications/editors/emacs/elisp-packages/manual-packages.nix]: https://github.com/NixOS/nixpkgs/pull/329249

File: pkgs/applications/editors/emacs/elisp-packages/manual-packages.nix [hunk 8]
```diff
@@ -69,7 +69,9 @@ in
 
   mu4e = callPackage ./manual-packages/mu4e { };
 
-  notdeft = callPackage ./manual-packages/notdeft { };
+  notdeft = callPackage ./manual-packages/notdeft {
+    inherit (pkgs) pkg-config tclap xapian;
+  };
```
Comment (jian-lin) [inline pkgs/applications/editors/emacs/elisp-packages/manual-packages.nix:74]: https://github.com/NixOS/nixpkgs/pull/329005#discussion_r1685882528

File: pkgs/applications/editors/emacs/elisp-packages/manual-packages/git-undo/default.nix [hunk 9]
```diff
@@ -15,6 +16,8 @@ melpaBuild {
     hash = "sha256-cVkK9EF6qQyVV3uVqnBEjF8e9nEx/8ixnM8PvxqCyYE=";
   };
 
+    passthru.updateScript = unstableGitUpdater { hardcodeZeroVersion = true; };
```
Comment (jian-lin) [inline pkgs/applications/editors/emacs/elisp-packages/manual-packages/git-undo/default.nix:19]: wrong indentation

**Review by jian-lin — COMMENTED** [review]
I think it is a good idea to ask their maintainers if they want an updateScript because it adds more maintenance burden to them.

**Comment by jian-lin** [comment]
> I think it is a good idea to ask their maintainers if they want an updateScript because it adds more maintenance burden to them.

CC @wuyoli because I cannot send you a review request using github UI

**Comment by AndersonTorres** [comment]
> adds more maintenance burden to them

This is already part of the maintenance burden.
Ryanbot auto-updates various packages without explicitly asking their maintainers - indeed it updates various packages without an explicit `updateScript`.

I would argue the autoupdate in fact relieves the burden, since the hash update is done by a machine.

**Comment by jian-lin** [comment]
> This is already part of the maintenance burden.
> Ryanbot auto-updates various packages without explicitly asking their maintainers - indeed it updates various packages without an explicit `updateScript`.

This is true if r-ryantm actually updates them without an updateScript.  However, I think none of these packages have ever been updated by r-ryantm before.

**Comment by AndersonTorres** [comment]
I splitted the packages I maintain in this PR, so that they can be reviewed quicker:

https://github.com/NixOS/nixpkgs/pull/329321

**Review by ibizaman — APPROVED** [review]

**Review by nessdoor — APPROVED** [review]

**Comment by AndersonTorres** [comment]
> This is true if r-ryantm actually updates them without an updateScript.

If a future release of Ryanbot becomes aware of these packages, this line of reasoning will fail.
Further, if a human person eventually updates this packages, the effect will be the same as a bot - the maintainer will be pinged.

Nonetheless I removed the package from @wuyoli, and I will remove the ones from @leungbk tomorrow.

**Review by jian-lin — COMMENTED** [review]
Could you update the commit message of `notdeft`?  Then I think this PR is in good shape.

**Comment by AndersonTorres** [comment]
Done.

The PR with the packages kept by leungbk was split.

**Review by jian-lin — APPROVED** [review]
LGTM

---

> If a future release of Ryanbot becomes aware of these packages, this line of reasoning will fail.
Further, if a human person eventually updates this packages, the effect will be the same as a bot - the maintainer will be pinged.

I do not agree but I do not want to discuss this further.
